contracts/vault.gno

```diff
@@ -0,0 +1,223 @@
+
+	// Send back any leftover tokens
+	amount0Parsed := subAmounts(amount0, amount0Used)
+	amount1Parsed := subAmounts(amount1, amount1Used)
+
+	if amount0Parsed.Cmp(uint256.NewUint(0)) > 0 {
+		transferByRegisterCall(vault.token0, caller, amount0Parsed.ToString())
+	}
+	if amount1Parsed.Cmp(uint256.NewUint(0)) > 0 {
+		transferByRegisterCall(vault.token1, caller, amount1Parsed.ToString())
+	}
+
+	std.Emit("Deposit", caller.String(), strconv.FormatUint(tokenId, 10), amount0Parsed.ToString(), amount1Parsed.ToString(), liquidity)
+}
+
+// DepositSingle deposits a single token into a vault
+// @param tokenId uint64 the vault id
+// @param token string the token to deposit
+// @param amount string the amount of token to deposit
+func DepositSingle(tokenId uint64, token string, amount string) {
+	vault := mustGetVault(tokenId)
+	assertVaultPaused(vault)
+
+	// TODO swap token to token0 or token1 if needed
+}
+
+// Withdraw withdraws tokens from a vault
+// @param tokenId uint64 the vault id
+// @param amount string the amount of liquidity to withdraw
+//
+// panic if the amount is greater than the caller's balance
+func Withdraw(tokenId uint64, amount string) {
+	vault := mustGetVault(tokenId)
+	caller := std.PrevRealm().Addr()
+	balance := mustGetBalance(vault, caller.String())
+	parsedAmount, err := uint256.FromDecimal(amount)
+	if err != nil {
+		panic(err)
+	}
+	if balance.Cmp(parsedAmount) < 0 {
+		panic("insufficient balance")
+	}
+
+	liquidityRatio := computeLiquidityRatio(tokenId, parsedAmount)
+	_, liquidity, _, _, amount0, amount1, _ := position.DecreaseLiquidity(tokenId, liquidityRatio, uint256.NewUint(0).ToString(), uint256.NewUint(0).ToString(), time.Now().Unix()+1, false)
+
+	// Transfer tokens to the caller
+	transferByRegisterCall(vault.token0, caller, amount0)
+	transferByRegisterCall(vault.token1, caller, amount1)
+
+	std.Emit("Withdraw", caller.String(), strconv.FormatUint(tokenId, 10), amount0, amount1, liquidity)
+}
+
+func Compound(tokenId uint64) {
+	vault := mustGetVault(tokenId)
+
+	// TODO implement
+}
+
+/**
+ * ADMIN FUNCTIONS
+ */
+
+// NewVault creates a new vault
+// @param token0 string the address of the first token
+// @param token1 string the address of the second token
+// @param fee uint32 the fee of the pool
+// @param tickLower int32 the lower tick of the pool
+// @param tickUpper int32 the upper tick of the pool
+// @param amount0 string the amount of token0 to deposit
+// @param amount1 string the amount of token1 to deposit
+// @param amount0Min string the minimum amount of token0 to deposit
+// @param amount1Min string the minimum amount of token1 to deposit
+// @param deadline int64 the deadline for the transaction
+func NewVault(
+	token0 string,
+	token1 string,
+	fee uint32,
+	tickLower int32,
+	tickUpper int32,
+	amount0 string, // *uint256.Uint
+	amount1 string, // *uint256.Uint
+	amount0Min string, // *uint256.Uint
+	amount1Min string, // *uint256.Uint
+	deadline int64,
+) {
```

Comment:
It seems need to validate that token0 and token1 are valid and distinct token addresses

Reply:
The check is already being done in the Mint function of the position realm so will eb a double check